lib/spack/spack/parser.py

```diff
@@ -278,7 +281,13 @@ def next_spec(
         initial_spec = initial_spec or spack.spec.Spec()
         root_spec = SpecNodeParser(self.ctx).parse(initial_spec)
         while True:
-            if self.ctx.accept(TokenType.DEPENDENCY):
+            if self.ctx.accept(TokenType.START_EDGE_PROPERTIES):
+                edge_properties = EdgeAttributeParser(self.ctx, self.literal_str).parse()
+                dependency = self._parse_node(root_spec)
+                edge_properties.setdefault("deptypes", ())
+                edge_properties.setdefault("virtuals", ())
+                root_spec._add_dependency(dependency, **edge_properties)
+            elif self.ctx.accept(TokenType.DEPENDENCY):
```

Comment:
I think this could be rewritten for more code reuse if you first check for either `DEPENDENCY` or `START_EDGE_PROPERTIES`.

```suggestion
            if self.ctx.accept(TokenType.Dependency) or self.ctx.accept(TokenType.START_EDGE_PROPERTIES):
                edge_properties = {}
                if self.ctx.token.kind == TokenType.START_EDGE_PROPERTIES:
                    edge_properties = EdgeAttributeParser(self.ctx, self.literal_str).parse()
                edge_properties.setdefault("deptypes", ())
                edge_properties.setdefault("virtuals", ())
```
And then the line 294/303 call to `_add_dependency` would change to:
```
root_spec._add_dependency(dependency, **edge_properties)
```

And then the entire `_parse_node` method that you had to add becomes redundant.

Reply:
I just reused `_parse_node` here (forgot to do it before)